share/wake/lib/core/json.wake

```diff
@@ -539,50 +539,190 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+# containing that double. This can be used to extract
+# a double from a JValue or in list operation chains
+# as a safe JValue to Double conversion.
+#
+# Examples:
+# ```
+#   jDouble (JDouble 3.14) = 3.14, Nil
+#   jDouble (JInteger 10) = Nil
+#   single json |> jField "height" |> jDouble = 1.75, Nil
+#   single json |> jField "name" |> jDouble = Nil
+# ```
 export def jDouble: JValue => List Double = match _
     JDouble d = single d
     _         = Nil
 
+# If `input` is a JDouble, updates the double using `editFn`.
+#
+# Parameters:
+#   - `editFn`: The function to update a double with
+#   - `input`: The JValue to be modified
+#
+# Examples:
+# ```
+#   editJDouble (_+.1.0) (JDouble 10.0) = JDouble 11.0
+#   editJDouble (_+.1.0) (JString "Bob") = JString "Bob"
+#   (editJField "tallestBuilding" $ editJDouble (\_ 1000.0)) statistics
+#   def grow: (person: JValue) => JValue = editJField "height" $ editJDouble (_+.0.02)
+# ```
 export def editJDouble (f: Double => Double): JValue => JValue = match _
     JDouble d = JDouble (f d)
     x         = x
 
+# If the input is a number, returns a singelton list
+# containing that number as a double. This can be used
+# to extract a number from a JValue or in list operation
+# chains as a safe JValue to number conversion.
+#
+# This works with both JDouble and JInteger. An integer
+# will be converted to a double if possible. If the
+# conversion is not possible because the integer is
+# out of range jNumber will return just Nil
+#
+# Examples:
+# ```
+#   jNumber (JDouble 3.14) = 3.14, Nil
+#   jNumber (JInteger 10) = 10.0, Nil
+#   jNumber (JString "Mal") = Nil
+#   jNumber (JInteger (factorial 100)) = Nil
+#   single json |> jField "height" |> jNumber = 1.75, Nil
+#   single json |> jField "age" |> jNumber = 28.0, Nil
+#   single json |> jField "name" |> jNumber = Nil
+# ```
 export def jNumber: JValue => List Double = match _
     JDouble x  = single x
     JInteger x = single x | mapPartial dint
     _          = Nil
 
+# If `input` is a number, updates the number using `editFn`.
+# This will edit both JDouble and JInteger. The JInteger will
+# be converted to a JDouble if the conversion from Integer
+# to Double succeeds. If the conversion from Integer to
+# double does not succeed the integer remains unedited.
+#
+# Parameters:
+#   - `editFn`: The function to update a number with
+#   - `input`: The JValue to be modified
+#
+# Examples:
+# ```
+#   editJNumber (_+.1.0) (JDouble 10.0) = JDouble 11.0
+#   editJNumber (_+.1.0) (JInteger 10) = JDouble 11.0
+#   editJNumber (_+.1.0) (JString "Bob") = JString "Bob"
+#   editJNumber (_+.1.0) (JInteger (factorial 100)) = JInteger (factorial 100)
+#   def newTallest: (stats: JValue) => JValue = editJField "tallestBuilding" $ editJNumber (\_ 1000.0)
+#   def grow: (person: JValue) => JValue = editJField "height" $ editJNumber (_+.0.02)
+# ```
 export def editJNumber (f: Double => Double): JValue => JValue = match _
```

Comment:
I know why you wanted getJNumber, but ... editJNumber changing the case seems kinda bad.

An alternative might be to change jDouble to doing this. Alternatively, we we could consider eliminating the JInteger case.

Reply:
> we we could consider eliminating the JInteger case

Actually, no. As a company, we need the ability to precisely encode 64-bit integers, so using JDouble for all numbers is a bad idea. SiFive needs the JInteger case. :-/

Reply:
We talked about this on zoom today, and we think this should just be the behaviour of jDouble. ie: remove jNumber.

Reply:
Yeah I wasn't happy with this but I'm also slightly unhappy with how json handles integers (or the fact that it doesn't rather) and different json libraries will handle it differently. This is why strings are often used for integers (because json handles this issue poorly).

Maybe we should do a brief audit of federation json libraries to make sure they all handle integers correctly? That would be the best of all worlds IMO.

Reply:
The only library that we care about for federation is the JSON => scala parser for the core generator, and it handles large integers correctly.

Reply:
The reason we decided on `jDouble` accepting both `JDouble` and `JInteger` is that when someone fills in a field, the parser has no way to know if that field is an integer or a double. So, currently `3` gets parsed as `JInteger` while `3.0` gets parsed as a `JDouble`. In order to make input configuration files work seamlessly, we really want `jDouble` to accept both formulations.

Reply:
(We could also consider allowing `jInteger` to accept a `JDouble` with a zero fraction)

Reply:
how do we feel about this

```
export def editJDouble (f: Double => Double): JValue => JValue = match _
    JDouble d = JDouble (f d)
    JInteger i = match (dint i)
        Some d =
            def result = f d
            def Pair whole frac = dmodf result
            if (dabs frac) ==. 0.0 then
                JInteger (whole)
            else
                JDouble result
        None = JInteger i
    x         = x
```

Reply:
Hmm. So, there are two things to talk about, I guess:
1. Do we want to convert back to JInteger from JInteger (`dmodf`)? I would lean towards: no. When we parsed the JSON, we didn't know that the correct type for this number was. However, since you just applied `editJDouble` to it, now we know.
2. What do we do if the integer overflows the double? I think leaving it as a JInteger is also probably wrong. I'd perhaps advocate for running `f` on `+/-inf`?

Reply:
```
export def editJDouble (f: Double => Double): JValue => JValue = match _
    JDouble d = JDouble (f d)
    JInteger i =
        dint i
        | getOrElse (if i < 0 then -.inf else inf)
        | f
        | JDouble
    x = x

export def editJInteger (f: Integer => Integer): JValue => JValue = match _
    JInteger i = JInteger (f i)
    JDouble d =
       def Pair whole frac = dmodf d
       if frac ==. 0.0 then JInteger (f whole) else JDouble d
    x = x
```

So these aren't perfectly symmetric, because `editJInteger` is more willing to 'give up' on the conversion. OTOH, if someone wrote `3.14` in the json, that's pretty good evidence it's not an integer.

Reply:
Yeah so locally I used your definition of `editJInteger` and updated `jInteger` and `jDouble` to match. editJDouble is really the only case I think we should drill down here.

I think your definition editJInteger is essentially correct here because it tries to convert to an integer whenever possible (per user intent) and only fails if there's a clear reason to fail (namely there was a non-zero fractional part).

editJDouble is more tricky I think. I can see both not editing in the case where i > 2^53 and your semantics as being desired. I would vote your definition if it wasn't for the fact that lots of functions behave really poorly when you pass them infinity.

I suppose editing on `inf` and `-inf` would fail more loudly in some sense that not editing would but thats just kind of a guess. I don't think there's a perfect answer here.

Reply:
I agree. All the proposals are just shades of bad. As you note, this problem comes from JSON itself, though, so there's not much we can do about it other than muddle through.

As another option, we could view `jInteger` and `jDouble` as 'cast' operators. In that scenario, perhaps `jInteger` is short for `jIntegerWith roundToNearest` and it explicitly coerces even fractional `Double`s. If we went that route, it becomes clear that `inf` is the right call for `jDouble`.

... depending on how far we want to take that idea, I wonder if we should even attempt a conversion from `JString`.

I mean, presumably, by the time you are applying `JInteger` / `jDouble` / `JString`, you've already selected the items you are looking for..? As you noted, people use JSON strings for integers, because JSON sucks. Maybe we should not be trying to force our types onto the typeless mess that is JSON?

Reply:
Ok I think I have a proposal I'm willing to stick to pending other thoughts...its basically your previous proposal.

In general I don't like giving user's too many implicit conversions and I especially don't like *forcing* implicit conversions on them where we don't have to. I generally think its better to error out if you get something unexpected than to try your best to chug along.

I agree that when the user asks for a `jInteger` they've told us that they want an integer but there's a missing detail needed  to do the right thing; we need to know the allowed encodings of that integer. Its important that the user be able to *reject* encodings that they didn't expect. Additionally as you noted with `roundToNearest` there isn't just one way to "decode" an integer from a double. Really that means we need a small zoo of options like `jIntegerOnly` `jIntegerWith` and `jIntegerFromString` and users might want combinations of this which they can achieve with `++` if desired...but that's all just too complicated; we're going to need some sensible defaults I think.

so `jString | mapPartial (intbase 10)` can be used for `jIntegerFromString` so I'm not super inclined to add that specifically right now (user's might also ant hex or octal). I think if the expected encoding is a string we can safely assume it will *only* be a string. But with numbers we probably have to be a bit more robust to mishaps, `jIntegerOnly` is just too strict in my opinion and `jDoubleOnly` is even worse. Those are not good defaults (I think we all agree on that)

for `jDouble` I think we need to accept Integers but not strings. We need this for the aforementioned case where someone manually writes `3` where they might have meant `3.0` (you can imagine a library doing this too). I think otherwise that's just an unacceptable "gotcha" to have to deal with. I think per the "implicit casting" idea I like using `inf` and `-inf`. I think that will give the best signal that there was a failure should that issue occur (and its not exactly likely to occur frankly). So I'd advocate for just this one very safe "implicit cast" here. I don't even think `jDoubleOnly` is needed.

for `jInteger` I think its more important to fail if the user supplies a non-zero fractional part than it is to attempt to chug along by rounding. I think we should accept doubles with a fractional part of zero though because a different json library might write an integer out as `3.0` instead of `3`. We can add `jIntegerWith` as well if the user wants more lenient behavior. This is less of a cast and more of just handling on specific unambiguous encoding of integers that might pop up in a "gotcha" kind of way.

for `editJDouble` I think I like your most recent proposal best, it matches on the `jDouble` semantics that I like and if someone says they want to set a field to a JDouble I think its sensible to set it to a Double; I'm not really inclined to take my "convert back to JInteger when possible" approach anymore. That's trying to be too clever and is akin to an implicit cast; if the user is really set on that behavior they can still add it themselves. This follows the rules that `setJFoo (f (getJFoo json)) json = editJFoo f json` which is the appropriate symmetry here; we can follow that blindly to determine the correct implementation of our `edit` functions I think (although it just boislt he problem down to what `setJFoo` should be. Here I claim that `setJDouble` should actually replace the JValue with a JDouble regardless of what was before.

for `editJInteger` we follow jInteger in the Double case to convert back only if the fractional part is zero but the final result is a `JInteger`. again I think this follows from `setJInteger (f (jInteger json)) json = editJInteger f json` where `setJInteger` always replaces the JValue with a `JInteger` and doesn't try to do anything too clever.

TL;DR I think this proposal (your code from a couple posts ago) is as good as we can get:

```
export def jDouble: JValue => List Double = match _
    JDouble d  = single d
    JInteger i =
        i
        | dint
        | getOrElse (if i < 0 then -.inf else inf)
        | single
    _         = Nil

export def editJDouble (f: Double => Double): JValue => JValue = match _
    JDouble d = JDouble (f d)
    JInteger i =
        dint i
        | getOrElse (if i < 0 then -.inf else inf)
        | f
        | JDouble
    x = x

export def jInteger: JValue => List Integer = match _
    JInteger i = single i
    JDouble d =
        def Pair whole frac = dmodf d
        if (dabs frac) ==. 0.0 then whole, Nil else Nil
    _          = Nil
 
export def editJInteger (f: Integer => Integer): JValue => JValue = match _
    JInteger i = JInteger (f i)
    JDouble d =
       def Pair whole frac = dmodf d
       if (dabs frac) ==. 0.0 then JInteger (f whole) else JDouble d
    x = x
```

I think it strikes the best possible balance between failing quickly/loudly and not having too many pointless gotchas. Plus while it might not seem symmetric at first glance the appropriate set/get/edit laws all correctly relate to each other (there is more than one solution but this is at least one of those solutions)

Reply:
> In general I don't like giving user's too many implicit conversions and I especially don't like _forcing_ implicit conversions on them where we don't have to. I generally think its better to error out if you get something unexpected than to try your best to chug along.

💯 

> `setJFoo (f (getJFoo json)) json = editJFoo f json` which is the appropriate symmetry here

💯 

> TL;DR I think this proposal (your code from a couple posts ago) is as good as we can get
> I think it strikes the best possible balance between failing quickly/loudly and not having too many pointless gotchas. Plus while it might not seem symmetric at first glance the appropriate set/get/edit laws all correctly relate to each other (there is more than one solution but this is at least one of those solutions)

Another way of making the delineation is that this approach fixes ONLY the mess we introduced (by splitting number into JInteger and JDouble).

Anyhow, sounds like we have consensus on the least bad alternative.

Reply:
I like it on the whole.  Two small questions, though: do we want to use `inf` or `nan` (no personal preference), and do we have to worry about floating-point inexactness in the `frac ==. 0.0` tests?

Reply:
Floating point in-exactness is primarily an issue when doing computations, not when reading/writing values intended to be integers in json (which is the case we're concerned with) so we should fail reasonably loudly if something winds up being inexact somewhere due to someone doing a computation with doubles. That said I did decide to include a `dabs` call to account for negative zero but thats the only real corner I think we can account for. conversion to double and back is exact (modulo infinity issues)

As for inf vs nan I was a bit more on the fence (but am no longer). I think we should use inf/-inf instead of NaN. NaN is more "infectious" and creates louder failures. NaN also causes strange control flow to occur when code isn't designed to handle it (weather this is good or bad I don't know, I think it "just is" personally). There are some rare cases where you might really want infinity as a result and you could get that back. For instance dividing by inf produces zero and all is right with the world if that's the behavior you wanted. I sort of think NaN vs infinity is essentially a wash. Maybe I lean towards inf a little bit because it does have *some* uses and I think those are more common than cases where `inf` gets accidentally covered up because it wasn't as infectious as NaN. I think that the number of cases where this will come up as an issue is very very low so the impact of this choice is somewhat lower as well.

Luckily we misunderstood how `dint` works and it breaks the tie for me. It turns out it never actually fails and produces a inf/-inf instead just like we wanted. So I think changing dint to not return an Option and instead always succeed, then making jDouble and friends always use dint to do the conversion would work well. So its simpler to implement with inf and otherwise a wash of an issue.